### PR TITLE
refactor: ES2021 features

### PR DESCRIPTION
**Pull request by ImRodry**
**Please describe the changes this PR makes and why it should be merged:**
This PR makes use of ES2021's features: nullish assignment, underscore numeric operator, and String.prototype.replaceAll! Here are the changes:
- Replaced (hopefully) all `if (prop) prop = X` and `if (typeof prop === 'undefined') prop = X` with `prop ??= X`. There may be situations where ||= might be more appropriate but I didn't find any when going through these, however please let me know if you do
- Added _ to separate big numbers (bigger than 999) but only on numbers with 000 at some point in their structure. If there is for example a 58395766 out there, I did not catch that. I also purposely left out Constants as these would become harder to compare to Discord's API docs
- Replaced some String.prototype.replace with String.prototype.replaceAll in Util methods
- Applied prettier on one of the test files because I messed up the formatting and it did that afterward, so might as well

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR **only** includes non-code changes, like changes to documentation, README, etc. (the changes are in the code but should not have any impact on anything on the library)

### CODE REVIEW

File: src/client/websocket/WebSocketManager.js [hunk 1]
```diff
@@ -19,7 +19,7 @@ const BeforeReadyWhitelist = [
 ];
 
 const UNRECOVERABLE_CLOSE_CODES = Object.keys(WSCodes).slice(1).map(Number);
-const UNRESUMABLE_CLOSE_CODES = [1000, 4006, 4007];
+const UNRESUMABLE_CLOSE_CODES = [1_000, 4_006, 4_007];
```
Comment (kyranet) [inline src/client/websocket/WebSocketManager.js:22]: Why not use `discord-api-types`? And same for the other codes in this and other files.

File: src/util/Util.js [hunk 2]
```diff
@@ -181,7 +181,7 @@ class Util extends null {
    * @returns {string}
    */
   static escapeCodeBlock(text) {
-    return text.replace(/```/g, '\\`\\`\\`');
+    return text.replaceAll(/```/g, '\\`\\`\\`');
```
Comment (kyranet) [inline src/util/Util.js:184]:
```suggestion
    return text.replaceAll('```', '\\`\\`\\`');
```

File: src/util/Util.js [hunk 3]
```diff
@@ -190,7 +190,7 @@ class Util extends null {
    * @returns {string}
    */
   static escapeInlineCode(text) {
-    return text.replace(/(?<=^|[^`])`(?=[^`]|$)/g, '\\`');
+    return text.replaceAll(/(?<=^|[^`])`(?=[^`]|$)/g, '\\`');
```
Comment (kyranet) [inline src/util/Util.js:193]: Unnecessary, please revert to `replace` as RegExp with global flag will go thru all instances.
```suggestion
    return text.replace(/(?<=^|[^`])`(?=[^`]|$)/g, '\\`');
```

File: src/util/Util.js [hunk 4]
```diff
@@ -200,12 +200,12 @@ class Util extends null {
    */
   static escapeItalic(text) {
     let i = 0;
-    text = text.replace(/(?<=^|[^*])\*([^*]|\*\*|$)/g, (_, match) => {
+    text = text.replaceAll(/(?<=^|[^*])\*([^*]|\*\*|$)/g, (_, match) => {
```
Comment (kyranet) [inline src/util/Util.js:203]: Unnecessary, please revert to `replace` as RegExp with global flag will go thru all instances.
```suggestion
    text = text.replace(/(?<=^|[^*])\*([^*]|\*\*|$)/g, (_, match) => {
```

File: src/util/Util.js [hunk 5]
```diff
@@ -200,12 +200,12 @@ class Util extends null {
...
       if (match === '**') return ++i % 2 ? `\\*${match}` : `${match}\\*`;
       return `\\*${match}`;
     });
     i = 0;
-    return text.replace(/(?<=^|[^_])_([^_]|__|$)/g, (_, match) => {
+    return text.replaceAll(/(?<=^|[^_])_([^_]|__|$)/g, (_, match) => {
```
Comment (kyranet) [inline src/util/Util.js:208]: Unnecessary, please revert to `replace` as RegExp with global flag will go thru all instances.
```suggestion
    return text.replace(/(?<=^|[^_])_([^_]|__|$)/g, (_, match) => {
```

File: src/util/Util.js [hunk 6]
```diff
@@ -218,7 +218,7 @@ class Util extends null {
    */
   static escapeBold(text) {
     let i = 0;
-    return text.replace(/\*\*(\*)?/g, (_, match) => {
+    return text.replaceAll(/\*\*(\*)?/g, (_, match) => {
```
Comment (kyranet) [inline src/util/Util.js:221]: Unnecessary, please revert to `replace` as RegExp with global flag will go thru all instances.
```suggestion
    return text.replace(/\*\*(\*)?/g, (_, match) => {
```

File: src/util/Util.js [hunk 7]
```diff
@@ -231,7 +231,7 @@ class Util extends null {
    */
   static escapeUnderline(text) {
     let i = 0;
-    return text.replace(/__(_)?/g, (_, match) => {
+    return text.replaceAll(/__(_)?/g, (_, match) => {
```
Comment (kyranet) [inline src/util/Util.js:234]: Unnecessary, please revert to `replace` as RegExp with global flag will go thru all instances.
```suggestion
    return text.replace(/__(_)?/g, (_, match) => {
```

File: src/util/Util.js [hunk 8]
```diff
@@ -243,7 +243,7 @@ class Util extends null {
    * @returns {string}
    */
   static escapeStrikethrough(text) {
-    return text.replace(/~~/g, '\\~\\~');
+    return text.replaceAll(/~~/g, '\\~\\~');
```
Comment (kyranet) [inline src/util/Util.js:246]:
```suggestion
    return text.replaceAll('~~', '\\~\\~');
```

File: src/util/Util.js [hunk 9]
```diff
@@ -252,7 +252,7 @@ class Util extends null {
    * @returns {string}
    */
   static escapeSpoiler(text) {
-    return text.replace(/\|\|/g, '\\|\\|');
+    return text.replaceAll(/\|\|/g, '\\|\\|');
```
Comment (kyranet) [inline src/util/Util.js:255]:
```suggestion
    return text.replaceAll('||', '\\|\\|');
```

File: src/util/Util.js [hunk 10]
```diff
@@ -577,7 +577,7 @@ class Util extends null {
    * @returns {string}
    */
   static removeMentions(str) {
-    return str.replace(/@/g, '@\u200b');
+    return str.replaceAll(/@/g, '@\u200b');
```
Comment (kyranet) [inline src/util/Util.js:580]:
```suggestion
    return str.replaceAll('@', '@\u200b');
```

File: src/util/Util.js [hunk 11]
```diff
@@ -621,7 +621,7 @@ class Util extends null {
    * @returns {string}
    */
   static cleanCodeBlockContent(text) {
-    return text.replace(/```/g, '`\u200b``');
+    return text.replaceAll(/```/g, '`\u200b``');
```
Comment (kyranet) [inline src/util/Util.js:624]:
```suggestion
    return text.replaceAll('```', '`\u200b``');
```

**Review by kyranet — DISMISSED** [review]

File: src/client/websocket/WebSocketManager.js (hunk 1 above)
Comment (ImRodry) [reply src/client/websocket/WebSocketManager.js:22]: Seems like that’s out of the scope of this PR and I’m also not very familiar with that package

File: src/client/Client.js [hunk 12]
```diff
@@ -181,7 +181,7 @@ class Client extends BaseClient {
       );
       this.sweepMessageInterval = setInterval(
         this.sweepMessages.bind(this),
-        this.options.messageSweepInterval * 1000,
+        this.options.messageSweepInterval * 1_000,
```
Comment (vladfrangu) [inline src/client/Client.js:184]: Am I the only one not a fan of using separators for such small numbers? I personally start using them from 10k+... @discordjs/the-big-3 thoughts?

File: src/structures/interfaces/TextBasedChannel.js [hunk 13]
```diff
@@ -297,7 +297,7 @@ class TextBasedChannel {
     if (Array.isArray(messages) || messages instanceof Collection) {
       let messageIds = messages instanceof Collection ? [...messages.keys()] : messages.map(m => m.id ?? m);
       if (filterOld) {
-        messageIds = messageIds.filter(id => Date.now() - SnowflakeUtil.deconstruct(id).timestamp < 1209600000);
+        messageIds = messageIds.filter(id => Date.now() - SnowflakeUtil.deconstruct(id).timestamp < 1_209_600_000);
```
Comment (vladfrangu) [inline src/structures/interfaces/TextBasedChannel.js:300]: Not for this PR but we should probably move this magic number to a const someday 👀

File: src/util/Options.js [hunk 14]
```diff
@@ -135,7 +135,6 @@ class Options extends null {
         version: 9,
       },
       http: {
-        agent: {},
```
Comment (vladfrangu) [inline src/util/Options.js:138]: Why was this removed?

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: src/client/Client.js (hunk 12 above)
Comment (ImRodry) [reply src/client/Client.js:184]: Personally I don’t think it looks very good on 1000 but other 4-digit numbers bigger than that are fine imo. I just ended up changing all because some places had 1000 and other numbers so it would seem odd for one to have it and not the others

Comment (kyranet) [reply src/client/Client.js:184]: My thoughts are the same.

File: src/structures/interfaces/TextBasedChannel.js (hunk 13 above)
Comment (ImRodry) [reply src/structures/interfaces/TextBasedChannel.js:300]: What even is this number?

File: src/util/Options.js (hunk 14 above)
Comment (ImRodry) [reply src/util/Options.js:138]: Probably some conflicts I came across when doing this. Thought I undid them all bug I’ll add this back

File: src/structures/interfaces/TextBasedChannel.js (hunk 13 above)
Comment (ckohen) [reply src/structures/interfaces/TextBasedChannel.js:300]: 2 weeks IIRC

Comment (ImRodry) [reply src/structures/interfaces/TextBasedChannel.js:300]: Could probably make this a constant inside this function then, since it’s not used anywhere and wouldn’t be out of the scope of the PR iirc

File: src/util/Options.js (hunk 14 above)
Comment (ImRodry) [reply src/util/Options.js]: Amended the commit

File: src/structures/interfaces/TextBasedChannel.js (hunk 13 above)
Comment (iCrawl) [reply src/structures/interfaces/TextBasedChannel.js:300]: Nah. We good.

File: src/structures/ClientUser.js [hunk 15]
```diff
@@ -25,8 +25,8 @@ class ClientUser extends User {
        * @type {?boolean}
        */
       this.mfaEnabled = typeof data.mfa_enabled === 'boolean' ? data.mfa_enabled : null;
-    } else if (typeof this.mfaEnabled === 'undefined') {
-      this.mfaEnabled = null;
+    } else {
+      this.mfaEnabled ??= null;
```
Comment (xiJulian) [inline src/structures/ClientUser.js:29]: i think that this should be left unchanged

Comment (ImRodry) [reply src/structures/ClientUser.js:29]: Why

Comment (xiJulian) [reply src/structures/ClientUser.js:29]: nvm seems like i didn't understand what does the nullish assignment actually do

Comment (ImRodry) [reply src/structures/ClientUser.js:29]: This only assigns the value if the property on the left-hand side is nullable

**Review by SpaceEEC — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

**Comment by ImRodry** [comment]
Cherry-picked the commit from #6544 since this PR was gonna touch that file anyway and it doesn't hurt to change this

File: src/client/websocket/WebSocketManager.js (hunk 1 above)
Comment (iCrawl) [reply src/client/websocket/WebSocketManager.js:22]: Then it's time to get familiar with it, I don't think this is out of scope at all, considering you just cherry-picked something completely out of scope yourself

Comment (ImRodry) [reply src/client/websocket/WebSocketManager.js:22]: Alright I’ll look into it in a bit, would also appreciate if you could suggest the changes and I’ll add the necessary imports above.

Comment (ImRodry) [reply src/client/websocket/WebSocketManager.js:22]: Addressed in the latest commit and also removed the out of scope commit, let me know if that works

**Review by iCrawl — APPROVED** [review]
